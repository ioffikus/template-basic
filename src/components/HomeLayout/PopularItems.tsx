import React from 'react';
import { RouterState } from 'connected-react-router';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Icon } from 'antd';
import ProductItems from 'src/components/ProductItems';

interface IOwnProps {}

interface IConnectedState {
  router: RouterState;
  ui: Alicanto.Models.UI;
  isRequesting: boolean;
  items: Alicanto.Models.CatalogItem[];
  hasNextPage: boolean;
}

type IProps = IOwnProps & IConnectedState;

interface IState {}

const mapStateToProps = (state: Store.IState) => ({
  router: state.router,
  ui: state.ui,
  isRequesting: state.popularItems.isRequesting,
  items: state.popularItems.content as Alicanto.Models.CatalogItem[],
  hasNextPage: false,
});

const WrappedPopularItems = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class PopularItems extends React.Component<IProps, IState> {
    render() {
      const popularBlockUI = this.props.ui.routes.home.popularBlock;
      return (
        <div className="root">
          <style jsx>{`
            .header {
              display: flex;
              flex-wrap: wrap;
              justify-content: space-between;
            }
            .root :global(.link) {
              margin-bottom: 30px;
              font-size: 16px;
              font-weight: 500;
              color: #062f49;
            }
          `}</style>
          <div className="header">
            <h2 className="app-title">{popularBlockUI.title.i18n.defaultText}</h2>
            <Link to="/catalog" className="link">
              {popularBlockUI.viewAll.i18n.defaultText} <Icon type="right" />
            </Link>
          </div>
          <ProductItems {...this.props} />
        </div>
      );
    }
  },
);

export default WrappedPopularItems;
