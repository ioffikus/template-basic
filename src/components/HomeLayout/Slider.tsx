import React from 'react';
import { connect } from 'react-redux';
import { Tag } from 'antd';
import Swiper from 'react-id-swiper';
import PreloadingImage from 'src/components/PreloadingImage';

interface IOwnProps {}

interface IConnectedState {
  ui: Alicanto.Models.UI;
}

type IProps = IOwnProps & IConnectedState;

interface IState {
  isMounted: boolean;
}

const mapStateToProps = (state: Store.IState) => ({
  ui: state.ui,
});

const WrappedSlider = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class Slider extends React.Component<IProps, IState> {
    state: IState = {
      isMounted: false,
    };

    swiperSettings = {
      slidesPerView: 1,
      loopAdditionalSlides: 0,
      loop: true,
      autoplay: {
        delay: 5000,
        disableOnInteraction: false,
      },
      navigation: {
        nextEl: '.swiper-button-next',
        prevEl: '.swiper-button-prev ',
      },
    };

    componentDidMount() {
      this.setState({
        isMounted: true,
      });
    }

    render() {
      const items = this.props.ui.routes.home.slider.items;
      return (
        <div className="root">
          <style jsx>{`
            .root {
              min-height: 512px;
              background-color: rgba(0, 0, 0, 0.2);
            }
            .root :global(.item) {
              min-height: 512px;
              display: flex;
              align-items: center;
              justify-content: center;
              position: relative;
              color: #ffffff;
              font-size: 48px;
              text-shadow: 0 0 4px 0 rgba(1, 2, 20, 0.78);
              font-weight: 500;
              line-height: 1.25;
            }
            .root :global(.image-spinner) {
              position: absolute;
              top: 0;
              left: 0;
              height: 100%;
              width: 100%;
              z-index: 1;
            }
            .root :global(.image-spinner .ant-spin-blur:after) {
              background: #000;
            }
            .root :global(.image) {
              height: 100%;
              width: 100%;
              object-fit: cover;
            }
            .inner {
              position: relative;
              z-index: 2;
              flex-basis: 920px;
              max-width: 920px;
              padding: 0 10px;
            }
            .root :global(.ant-tag) {
              margin-bottom: 30px;
              height: auto;
              font-size: 24px;
              font-weight: bold;
              text-transform: uppercase;
              cursor: default;
              background: transparent;
              border-color: #fff;
              color: inherit;
              line-height: 1.25;
            }
            .title {
              font-size: 48px;
              color: inherit;
              line-height: 1.25;
            }
            .text {
              margin-bottom: 40px;
              text-transform: uppercase;
            }
          `}</style>
          {this.state.isMounted && (
            <Swiper {...this.swiperSettings}>
              {Object.keys(items).map(key => {
                const item = items[key];
                const image = item.image.settings.link;
                const title = item.header.i18n.defaultText;
                const isTitle = item.header.settings.isVisible;
                const text = item.text.i18n.defaultText;
                const isText = item.text.settings.isVisible;
                const price = item.price.i18n.defaultText;
                const isPrice = item.price.settings.isVisible;
                const tag = item.tag.i18n.defaultText;
                const isTag = item.tag.settings.isVisible;
                return (
                  <div key={key} className="item">
                    <PreloadingImage src={image} imageClass="image" wrapperClass="image-spinner" />
                    <div className="inner">
                      {isTag && <Tag>{tag}</Tag>}
                      {isTitle && <h1 className="title">{title}</h1>}
                      {isText && <p className="text">{text}</p>}
                      {isPrice && <div className="price">{price}</div>}
                    </div>
                  </div>
                );
              })}
            </Swiper>
          )}
        </div>
      );
    }
  },
);

export default WrappedSlider;
