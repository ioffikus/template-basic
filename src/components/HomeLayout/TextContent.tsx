import React from 'react';
import { connect } from 'react-redux';
import { Card, Row, Col } from 'antd';

interface IOwnProps {}

interface IConnectedState {
  ui: Alicanto.Models.UI;
}

type IProps = IOwnProps & IConnectedState;

interface IState {}

const mapStateToProps = (state: Store.IState) => ({
  ui: state.ui,
});

const WrappedTextContent = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class TextContent extends React.Component<IProps, IState> {
    render() {
      const textUI = this.props.ui.routes.home.textBlock;
      return (
        <div className="root">
          <div className="app-section">
            <h2 className="app-title">{textUI.title.i18n.defaultText}</h2>
            <Row type="flex">
              <Col span={12}>
                <Card title={textUI.subtitle.i18n.defaultText} noHovering={true} bordered={false} className="app-cart">
                  <p className="app-text">{textUI.text.i18n.defaultText}</p>
                </Card>
              </Col>
              <Col span={12}>
                <Card title={textUI.subtitle.i18n.defaultText} noHovering={true} bordered={false} className="app-cart">
                  <p className="app-text">{textUI.text.i18n.defaultText}</p>
                </Card>
              </Col>
            </Row>
          </div>
          <div className="app-section">
            <h2 className="app-title">{textUI.title.i18n.defaultText}</h2>
            <Row type="flex">
              <Col span={8}>
                <Card title={textUI.subtitle.i18n.defaultText} noHovering={true} bordered={false} className="app-cart">
                  <p className="app-text">{textUI.text.i18n.defaultText}</p>
                </Card>
              </Col>
              <Col span={8}>
                <Card
                  title={textUI.subtitle.i18n.defaultText}
                  noHovering={true}
                  bordered={false}
                  className="app-cart app-text-center"
                >
                  <p className="app-text">{textUI.text.i18n.defaultText}</p>
                </Card>
              </Col>
              <Col span={8}>
                <Card
                  title={textUI.subtitle.i18n.defaultText}
                  noHovering={true}
                  bordered={false}
                  className="app-cart app-text-right"
                >
                  <p className="app-text">{textUI.text.i18n.defaultText}</p>
                </Card>
              </Col>
            </Row>
          </div>
        </div>
      );
    }
  },
);

export default WrappedTextContent;
