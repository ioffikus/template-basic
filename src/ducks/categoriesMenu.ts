import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { GET } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'CATEGORIES_MENU';
const NS = `${REDUCER}__`;

interface ICategoriesMenuResponse extends Alicanto.API.GET_CATALOG_GROUPS_SLUG_RESPONSE {
  content: Alicanto.Models.Category[];
}

export interface ICategoriesMenuState extends ICategoriesMenuResponse, Store.IReduxCommonState {}

export const initialState: ICategoriesMenuState = {
  slug: '',
  type: '',
  content: [],
  isRequesting: false,
  isRequestError: false,
};

const reducer = createReducer<ICategoriesMenuState>({}, initialState);

const callSuccess = createAction<ICategoriesMenuResponse>(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, (state, { slug, type, content }) => {
  return {
    ...state,
    slug,
    type,
    content: [...content],
    isRequestError: false,
    isRequesting: false,
  };
});

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  slug: '',
  type: '',
  content: [],
  isRequesting: false,
  isRequestError: true,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequesting: true,
}));

export const getCategoriesMenu = (abortToken: string): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    GET(
      {
        url: EndpointsConfig.CATEGORIES_MENU,
      },
      xShopHeaders,
      abortToken,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as ICategoriesMenuResponse;
        dispatch(callSuccess(data));
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

export default reducer;
