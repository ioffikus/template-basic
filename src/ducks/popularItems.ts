import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { GET } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'POPULAR_ITEMS';
const NS = `${REDUCER}__`;

interface IPopularItemsResponse extends Alicanto.API.GET_CATALOG_GROUPS_SLUG_RESPONSE {
  content: Alicanto.Models.CatalogItem[];
}

export interface IPopularItemsState extends Alicanto.API.GET_CATALOG_GROUPS_SLUG_RESPONSE, Store.IReduxCommonState {}

export const initialState: IPopularItemsState = {
  slug: '',
  type: '',
  content: [],
  isRequesting: false,
  isRequestError: false,
};

const reducer = createReducer<IPopularItemsState>({}, initialState);

const callSuccess = createAction<IPopularItemsResponse>(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, (state, { slug, type, content }) => ({
  ...state,
  slug,
  type,
  content: [...content],
  isRequestError: false,
  isRequesting: false,
}));

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  slug: '',
  type: '',
  content: [],
  isRequesting: false,
  isRequestError: true,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  content: [],
  isRequesting: true,
}));

export const getPopularItems = (abortToken: string): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    GET(
      {
        url: EndpointsConfig.POPULAR_ITEMS,
      },
      xShopHeaders,
      abortToken,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as IPopularItemsResponse;
        dispatch(callSuccess(data));
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

export default reducer;
