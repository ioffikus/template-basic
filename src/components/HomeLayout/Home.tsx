import React from 'react';
import Slider from './Slider';
import TextContent from './TextContent';
import PopularItems from './PopularItems';

class Home extends React.Component<{}, {}> {
  render() {
    return (
      <div className="root">
        <style jsx>{`
          .root {
            width: 100%;
          }
        `}</style>
        <Slider />
        <div className="app-container">
          <TextContent />
          <div className="app-section">
            <PopularItems />
          </div>
        </div>
      </div>
    );
  }
}

export default Home;
