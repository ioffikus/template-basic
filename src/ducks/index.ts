// instead this, we have router in store via connected-react-router
// import { routerReducer as routing, RouterState } from 'react-router-redux';
import { combineReducers } from 'redux';
import auth from './auth';
import cart from './cart';
import cartPaidItems from './cartPaidItems';
import catalogQuery from './catalogQuery';
import catalog from './catalog';
import categories from './categories';
import categoriesMenu from './categoriesMenu';
import filter from './filter';
import locale from './locale';
import meta from './meta';
import page from './page';
import product from './product';
import progress from './progress';
import popularItems from './popularItems';
import recommendItems from './recommendItems';
import routes from './routes';
import ui from './ui';
import user from './user';
import userDelivery from './userDelivery';
import userOrders from './userOrders';

export default combineReducers<Store.IState>({
  ui,
  auth,
  user,
  categories,
  categoriesMenu,
  catalogQuery,
  catalog,
  product,
  progress,
  routes,
  meta,
  filter,
  cart,
  cartPaidItems,
  userDelivery,
  userOrders,
  popularItems,
  recommendItems,
  page,
  locale,
});
