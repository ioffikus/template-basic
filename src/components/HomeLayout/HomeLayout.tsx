import React from 'react';
import { Layout } from 'antd';
import dispatch from 'src/store/dispatch';
import { getCategoriesMenu } from 'src/ducks/categoriesMenu';
import { getPopularItems } from 'src/ducks/popularItems';
import SiderMenu from 'src/components/SiderMenu';
import CategoriesMenu from 'src/components/CategoriesMenu';
import Home from './Home';

interface IProps {}

class HomeLayout extends React.Component<IProps, {}> {
  static fetchData: Store.IFetchDataFn = abortToken => {
    return Promise.all([dispatch(getCategoriesMenu(abortToken)), dispatch(getPopularItems(abortToken))]);
  };
  render() {
    return (
      <Layout>
        <SiderMenu type="categories">
          <CategoriesMenu />
        </SiderMenu>
        <Layout.Content>
          <Home />
        </Layout.Content>
      </Layout>
    );
  }
}

export default HomeLayout;
