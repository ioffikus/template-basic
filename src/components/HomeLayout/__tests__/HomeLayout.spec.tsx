import React from 'react';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import nanoid from 'nanoid';
import { setupDispatch } from 'src/store/dispatch';
import { getCategoriesMenu } from 'src/ducks/categoriesMenu';
import { getPopularItems } from 'src/ducks/popularItems';
import HomeLayout from '../HomeLayout';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
jest.mock('src/ducks/categoriesMenu', () => {
  return {
    getCategoriesMenu: jest.fn(() => {
      return () => Promise.resolve();
    }),
  };
});
jest.mock('src/ducks/popularItems', () => {
  return {
    getPopularItems: jest.fn(() => {
      return () => Promise.resolve();
    }),
  };
});

describe(null, () => {
  it('fetchData()', () => {
    const store = mockStore({});
    setupDispatch({ store });
    const abortToken = nanoid();
    return HomeLayout.fetchData(abortToken).then(data => {
      expect(getCategoriesMenu).toBeCalledWith(abortToken);
      expect(getPopularItems).toBeCalledWith(abortToken);
    });
  });
  it('render()', () => {
    const store = mockStore({});
    const wrapper = shallow(<HomeLayout />, { context: { store } });
    expect(wrapper).toMatchSnapshot();
  });
});
