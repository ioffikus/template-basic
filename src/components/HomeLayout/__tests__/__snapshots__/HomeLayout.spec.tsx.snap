// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() 1`] = `
<Adapter>
  <SiderMenu
    type="categories"
  >
    <Connect(CategoriesMenu) />
  </SiderMenu>
  <Adapter>
    <Home />
  </Adapter>
</Adapter>
`;
